File: C++/fig4.cpp
#include<iostream>
#include<fstream>
#include "NSFEM.h"
using namespace NSFEM;

main() {
    int i;
    double x,y;
    std::ofstream f;

    init("../cylinder/node.txt",
         "../cylinder/element.txt",
         "../cylinder/stream_fix.txt",
         "../cylinder/vortex_fix.txt",
         "../cylinder/vortex_wall.txt");
    // Karman vortex appears behind the cylinder
    i = run(0.5, 0.001, 0.003, 440);
    std::cout << i << std::endl;

    f.open("fig4uv.txt");
    for(i=0; i<element.nrows(); i++) {
        elem_center(x, y, i);
        f << x << ' ';
        f << y << ' ';
        f << velocity[i][0] << ' ';
        f << velocity[i][1] << '\n';
    }
    f.close();

    f.open("fig4ps.txt");
    for(i=0; i<node.nrows(); i++) {
        f << node[i][0] << ' ';
        f << node[i][1] << ' ';
        f << stream[i] << '\n';
    }
    f.close();

    f.open("fig4om.txt");
    for(i=0; i<node.nrows(); i++) {
        f << node[i][0] << ' ';
        f << node[i][1] << ' ';
        f << vortex[i] << '\n';
    }
    f.close();
}

File: fig4.py
import numpy as np
import matplotlib.pyplot as plt
from NSFEM import NSFEM
from contour import contour

ns = NSFEM('cylinder/node.txt',
           'cylinder/element.txt',
           'cylinder/stream_fix.txt',
           'cylinder/vortex_fix.txt',
           'cylinder/vortex_wall.txt')
region = np.loadtxt('cylinder/region.txt')

# Karman vortex appears behind the cylinder
n = ns.run(0.5, 0.001, 0.003, maxit=440)
print(n)

plt.figure(figsize=(5,8.2))

x,y = ns.node.T
X,Y = region.T

plt.subplot(311)
plt.axis('equal')
plt.axis([0,20,0,10])
ns.plot_velocity()
plt.plot([0,20,20,0,0], [0,0,10,10,0], 'k')
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.subplot(312)
plt.axis('equal')
plt.axis([0,20,0,10])
contour(x, y, ns.stream, 24, exclude=region, cmap='jet')
plt.plot([0,20,20,0,0], [0,0,10,10,0], 'k')
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.subplot(313)
plt.axis('equal')
plt.axis([0,20,0,10])
contour(x, y, ns.vortex, 32, exclude=region, cmap='jet')
plt.plot([0,20,20,0,0], [0,0,10,10,0], 'k')
plt.plot(X, Y, 'k')
plt.box('off')
plt.xticks(())
plt.yticks(())

plt.tight_layout()
plt.savefig('fig4.eps')
plt.show()
